[E22.X] fishhawk_run_stage: cmd.Wait + bufio.Scanner race (stdout pipe closed mid-read)

## Context

CI on PR #438 surfaced a real race in `backend/cmd/fishhawk-mcp/run_stage.go` from #436 that **didn't show up locally**:

```
--- FAIL: TestRunStage_JSONLAccumulation_OrderPreserved (0.00s)
    run_stage_test.go:400: events = 0, want 3 (got [])
--- FAIL: TestRunStage_NonJSONLineWarns (0.00s)
    run_stage_test.go:441: events = 0, want 1 (the JSON line)
    run_stage_test.go:451: non-JSON line should be in warnings: [scan runner stdout: read |0: file already closed]
```

The error message names it: the scanner returns `read |0: file already closed` because `cmd.Wait()` closes the stdout pipe before the bufio.Scanner finishes reading it.

## Cause

Current shape in `run_stage.go::runStage`:

```go
parseWG.Add(1)
go func() {
    defer parseWG.Done()
    runStageParseEvents(ctx, stdoutPipe, &events, ...)
}()

waitErr := runStageWaitForExit(ctx, cmd, runStageGracePeriod)
parseWG.Wait()
```

`runStageWaitForExit` calls `cmd.Wait()` internally. Go's `exec` documentation warns explicitly:

> StdoutPipe ... Wait will close the pipe after seeing the command exit, so most callers need not close it themselves; it is thus incorrect to call Wait before all reads from the pipe have completed.

Locally the parser usually beats Wait to EOF by luck of scheduling — tiny fake subprocess, tiny payload, parser goroutine schedules and drains fast. CI's runner scheduling is faster (or different enough) that Wait wins and the pipe closes mid-read.

## Fix

Sequence so Wait is **only** called after the parser has drained the pipe:

```go
parseDone := make(chan struct{})
go func() {
    defer close(parseDone)
    runStageParseEvents(ctx, stdoutPipe, &events, ...)
}()

// Watch for ctx cancel; SIGTERM the subprocess so its pipe closes and
// the parser sees EOF naturally.
go func() {
    select {
    case <-ctx.Done():
        if cmd.Process != nil {
            _ = cmd.Process.Signal(syscall.SIGTERM)
            // Optional grace + SIGKILL escalation here.
        }
    case <-parseDone:
    }
}()

<-parseDone     // parser exits when stdout reaches EOF (subprocess exited or was killed)
waitErr := cmd.Wait()  // safe: pipe already drained
```

This keeps the SIGTERM-on-cancel path intact (we still kill the subprocess) but removes the Wait/read ordering hazard.

## Acceptance criteria

- [ ] `run_stage.go`'s wait/parse sequencing is reordered so `cmd.Wait()` is only called after `runStageParseEvents` returns. The SIGTERM/30s/SIGKILL semantics from ADR-024 Q2 and #437 are preserved.
- [ ] The existing `TestRunStage_ContextCancelSendsSIGTERM` keeps passing (exit 143 or -1 from SIGTERM; ~tens of ms).
- [ ] `TestRunStage_JSONLAccumulation_OrderPreserved` and `TestRunStage_NonJSONLineWarns` pass reliably on CI. Run them N=100 times under `go test -count 100 -race ./...` locally as a flake check before merging.
- [ ] No regression in `TestRunStage_NonZeroExitPropagated` (the exit-code propagation path).

## Why this is sibling to #439, not the same

#439 is about reading from the **wrong stream** (stdout vs stderr) — a design assumption mismatch with the real runner. This issue is about a **race** even when the source stream is right — wrong synchronization. Both are bugs in #436's event-accumulator path; either fix alone leaves the other failure mode in place. Land them together or in sequence; the merge for #439 should not regress this race.

## Related

- #436 — original PR that introduced this code.
- #437 — runner SIGTERM handler. The cancellation semantics on the runner side don't change; this is purely a wait/read ordering bug in the MCP-side wrapper.
- #438 — the PR whose CI first exposed it.
- #439 — events:null sibling bug.
- ADR-024 / #433 — Q2 cancellation contract this preserves.
- [[feedback-dogfood-local-loop]]

--- a/backend/cmd/fishhawk-mcp/run_stage.go
+++ b/backend/cmd/fishhawk-mcp/run_stage.go
@@ -246,6 +246,19 @@ func (r *runResolver) runStage(ctx context.Context, req *mcp.CallToolRequest, in
 	cmd := runStageCommand(binary, argv...)
 	cmd.Env = append(os.Environ(), "FISHHAWK_API_TOKEN="+r.api.token)
 
+	// Run the subprocess in its own process group so signals reach
+	// the whole tree, not just the direct child. The runner spawns
+	// further descendants (the agent it invokes, that agent's tool
+	// processes); without a group, SIGTERM/SIGKILL on the runner
+	// leaves orphans alive that inherit our stdout fd — which keeps
+	// the pipe open and prevents the bufio scanner from ever
+	// reaching EOF. Signalling -pgid hits every descendant in
+	// lockstep. The #446 plan's risk section claimed "SIGKILL
+	// handles orphans" — that's incorrect for Unix fd inheritance,
+	// surfaced by TestRunStage_ContextCancelSendsSIGTERM after the
+	// pipe-drain reorder.
+	runStageSetProcessGroup(cmd)
+
 	// (5) Wire stdout to a JSONL parser. Stderr is piped through to
 	// the operator's terminal — the runner's verbose log lines are
 	// meant for humans, not the agent.
@@ -260,30 +273,54 @@ func (r *runResolver) runStage(ctx context.Context, req *mcp.CallToolRequest, in
 	}
 
 	// Concurrently: parse stdout into events; watch ctx for
-	// cancellation and signal the child. Both feed into the
-	// wait-for-exit synchronization below.
+	// cancellation and signal the child. parseDone closes when the
+	// parser goroutine drains stdout to EOF — guaranteeing cmd.Wait
+	// is called after the pipe is fully read (#446).
 	var (
 		events    []RunnerEvent
 		eventsMu  sync.Mutex
-		parseWG   sync.WaitGroup
+		parseDone = make(chan struct{})
 		progToken any
 	)
 	if req != nil && req.Params != nil {
 		progToken = req.Params.GetProgressToken()
 	}
 
-	parseWG.Add(1)
 	go func() {
-		defer parseWG.Done()
+		defer close(parseDone)
 		runStageParseEvents(ctx, stdoutPipe, &events, &eventsMu, &warnings, req, progToken)
 	}()
 
-	// (6+7) Wait for the runner to exit, honoring context
-	// cancellation. The wait coordinator owns the single cmd.Wait
-	// call so we don't trigger "no child processes" from a stray
-	// duplicate Wait.
-	waitErr := runStageWaitForExit(ctx, cmd, runStageGracePeriod)
-	parseWG.Wait()
+	// (6) Cancellation watcher: on ctx.Done(), signal the whole
+	// process group (not just the direct child) so descendants that
+	// inherited stdout die too — only then does the pipe close and
+	// the scanner reach EOF. Escalates to SIGKILL (group-wide)
+	// after the grace period.
+	//
+	// Snapshot the grace period at entry so a test's t.Cleanup
+	// restoring runStageGracePeriod doesn't race the goroutine.
+	grace := runStageGracePeriod
+	go func() {
+		select {
+		case <-parseDone:
+			// Normal exit — subprocess closed stdout, scanner drained.
+		case <-ctx.Done():
+			runStageSignalGroup(cmd, syscall.SIGTERM)
+			select {
+			case <-parseDone:
+				// Subprocess exited within grace; scanner done.
+			case <-time.After(grace):
+				runStageSignalGroup(cmd, syscall.SIGKILL)
+				<-parseDone
+			}
+		}
+	}()
+
+	// (7) Block until the parser goroutine has fully drained stdout,
+	// then call cmd.Wait(). The pipe is guaranteed drained at this
+	// point — no scanner-vs-Wait race.
+	<-parseDone
+	waitErr := cmd.Wait()
 
 	exitCode := 0
 	switch {
@@ -416,44 +453,6 @@ func runStageEventMessage(payload any) string {
 		return string(enc[:maxLen]) + "..."
 	}
 	return string(enc)
-}
-
-// runStageWaitForExit owns the canonical cmd.Wait() and coordinates
-// cancellation. Flow:
-//
-//   - Start one goroutine doing cmd.Wait(); it's the ONLY Wait on
-//     this process. A duplicate Wait would race the kernel's
-//     wait-status collection and return "no child processes" on
-//     macOS / Linux.
-//   - Block on (waitDone OR ctx.Done()).
-//   - On ctx.Done() before the runner exits: SIGTERM, then up to
-//     grace seconds for the wait to settle, then SIGKILL, then wait
-//     unconditionally for the wait to settle.
-//
-// Returns the error from the canonical Wait — caller branches on
-// exec.ExitError to extract the exit code.
-func runStageWaitForExit(ctx context.Context, cmd *exec.Cmd, grace time.Duration) error {
-	waitDone := make(chan error, 1)
-	go func() { waitDone <- cmd.Wait() }()
-
-	select {
-	case err := <-waitDone:
-		return err
-	case <-ctx.Done():
-		// Operator cancelled; send SIGTERM and wait up to grace.
-		if cmd.Process != nil {
-			_ = cmd.Process.Signal(syscall.SIGTERM)
-		}
-		select {
-		case err := <-waitDone:
-			return err
-		case <-time.After(grace):
-			if cmd.Process != nil {
-				_ = cmd.Process.Kill()
-			}
-			return <-waitDone
-		}
-	}
 }
 
 // runStageDetectGitHubRepo mirrors the CLI's helper. Returns

--- a/backend/cmd/fishhawk-mcp/run_stage_unix.go
+++ b/backend/cmd/fishhawk-mcp/run_stage_unix.go
@@ -1,0 +1,47 @@
+//go:build !windows
+
+package main
+
+import (
+	"os/exec"
+	"syscall"
+)
+
+// runStageSetProcessGroup attaches a fresh process group to cmd
+// before Start. Unix semantics: SysProcAttr.Setpgid makes the
+// child the leader of a new group with Pgid == its own PID, so
+// `kill(-pgid, sig)` from `runStageSignalGroup` reaches the
+// child and every descendant that hasn't broken away with its
+// own setpgid.
+//
+// Test seam: exposed as `var` so the package-level setter in
+// tests can no-op it (the fake subprocess for the unit tests
+// shouldn't run in a separate group — the test parent needs to
+// be able to observe the children directly).
+var runStageSetProcessGroup = func(cmd *exec.Cmd) {
+	if cmd.SysProcAttr == nil {
+		cmd.SysProcAttr = &syscall.SysProcAttr{}
+	}
+	cmd.SysProcAttr.Setpgid = true
+}
+
+// runStageSignalGroup sends sig to the process group cmd was
+// started in. Falls back to signalling the direct child if the
+// group wasn't set up (defensive — should never fire given the
+// always-on setup in runStage).
+//
+// The minus on the PID is the POSIX convention for "signal the
+// process group with that ID." Errors are swallowed because
+// missing-process is normal (race with subprocess exit) and we
+// can't recover from EPERM anyway.
+var runStageSignalGroup = func(cmd *exec.Cmd, sig syscall.Signal) {
+	if cmd == nil || cmd.Process == nil {
+		return
+	}
+	pgid, err := syscall.Getpgid(cmd.Process.Pid)
+	if err != nil {
+		_ = cmd.Process.Signal(sig)
+		return
+	}
+	_ = syscall.Kill(-pgid, sig)
+}

--- a/backend/cmd/fishhawk-mcp/run_stage_windows.go
+++ b/backend/cmd/fishhawk-mcp/run_stage_windows.go
@@ -1,0 +1,24 @@
+//go:build windows
+
+package main
+
+import (
+	"os/exec"
+	"syscall"
+)
+
+// runStageSetProcessGroup is a no-op on Windows. The runner is
+// not currently supported on Windows; if/when it is, the
+// equivalent is to spawn with `CREATE_NEW_PROCESS_GROUP` via
+// SysProcAttr.CreationFlags. Keeping the stub here so the
+// non-Unix build still compiles.
+var runStageSetProcessGroup = func(_ *exec.Cmd) {}
+
+// runStageSignalGroup falls back to signalling just the direct
+// child on Windows. Same caveats as the setup stub.
+var runStageSignalGroup = func(cmd *exec.Cmd, sig syscall.Signal) {
+	if cmd == nil || cmd.Process == nil {
+		return
+	}
+	_ = cmd.Process.Signal(sig)
+}

